### PR TITLE
only load app if necessary

### PR DESCRIPTION
**Pull request by snopoke**
##### SUMMARY
The changes in PR (https://github.com/dimagi/commcare-hq/pull/27587) result in loading the app on every request (regardless of cached values). This puts a lot of extra load on couchdb. This PR reverts to the previous mode of only loading the app if necessary.

This will still load the app a lot more than before since we are now caching on `app_id, build_profile_id` instead of just `app_id`. But I don't see a way to avoid that.

### CODE REVIEW

File: corehq/apps/app_manager/models.py [hunk 1]
```diff
@@ -5759,18 +5759,17 @@ class Meta(object):
...
-            'app_version': LATEST_APP_VALUE,
-        })
+        model = cls.by_app_id(app.domain, app.master_id)
         model._app = app
         return model
 
     @classmethod
     def by_app_id(cls, domain, app_id):
-        app = get_app(domain, app_id, latest=True, target='release')
-        assert app_id == app.master_id, "this class doesn't handle copy app ids"
```
Comment (sravfeyn) [inline corehq/apps/app_manager/models.py:5772]: This validation isn't necessary, is it?

Comment (sravfeyn) [reply corehq/apps/app_manager/models.py:5772]: Ah, you moved it below.

File: corehq/apps/app_manager/models.py [hunk 2]
```diff
@@ -5779,6 +5778,14 @@ def save(self, force_insert=False, force_update=False, using=DEFAULT_DB_ALIAS, u
             force_insert=force_insert, force_update=force_update, using=using, update_fields=update_fields
         )
 
+    @property
+    def app(self):
+        if not self._app:
+            app = get_app(self.domain, self.app_id, latest=True, target='release')
+            assert self.app_id == app.master_id, "this class doesn't handle copy app ids"
```
Comment (sravfeyn) [inline corehq/apps/app_manager/models.py:5785]: If we can't quickcache `get_app` since it's hard to bust the cache, we could pull out a separate method such as `is_master_app` and quickcache it. It would be easy to bust the cache, we could avoid fetching the app on every request that way.

Comment (sravfeyn) [reply corehq/apps/app_manager/models.py:5785]: It looks like more things are needed though like `is_released and version`, still possible to cache these?

File: corehq/apps/app_manager/models.py (hunk 1 above)
Comment (snopoke) [reply corehq/apps/app_manager/models.py:5772]: it's moved to where the app is fetched

File: corehq/apps/app_manager/models.py (hunk 2 above)
Comment (snopoke) [reply corehq/apps/app_manager/models.py:5785]: I'm not sure I follow, you're saying just cache the assertion? That's not really the issue, the app was being loaded for every request instead of only requests that don't have a cached value.

File: corehq/apps/app_manager/models.py [hunk 3]
```diff
@@ -5759,18 +5759,17 @@ class Meta(object):
...
+        model = cls.by_app_id(app.domain, app.master_id)
         model._app = app
         return model
 
     @classmethod
     def by_app_id(cls, domain, app_id):
-        app = get_app(domain, app_id, latest=True, target='release')
-        assert app_id == app.master_id, "this class doesn't handle copy app ids"
-        return cls.by_app(app)
+        model, created = cls.objects.get_or_create(app_id=app_id, domain=domain, defaults={
```
Comment (mkangia) [inline corehq/apps/app_manager/models.py:5768]: Hey @snopoke,
So now we are going to just create the record, without checking if the app id actually belongs to the master app?

And plus, now we don't raise the Assertion Error so it won't get caught in the view like earlier since we have moved the check to later but fail when calling `get_latest_app_version` which will raise an uncaught assertion error?

Comment (snopoke) [reply corehq/apps/app_manager/models.py:5768]: good point about not catching the assertion in the view

Yes, this will create an object in the DB even it's not a 'master' app ID but that's not very likely. I don't think we've ever seen that in the wild (search sentry for "Received an invalid heartbeat request").

Comment (mkangia) [reply corehq/apps/app_manager/models.py:5768]: I think its fine if we want to just create it now, but would be nice to just delete it if that is the case when fetching the app?

And how do we avoid fetching the app in the request now? looks like we are just delaying it?

Comment (snopoke) [reply corehq/apps/app_manager/models.py:5768]: catch assertions: 9754f17d318

File: corehq/apps/ota/views.py [hunk 4]
```diff
@@ -302,19 +302,26 @@ def heartbeat(request, domain, app_build_id):
...
     try:
+        version_info = _get_version_info(config)
         # mobile will send master app_id
-        config = GlobalAppConfig.by_app_id(domain, app_id)
     except (Http404, AssertionError):
         # If it's not a valid master app id, find it by talking to couch
         notify_exception(request, 'Received an invalid heartbeat request')
         app = get_app_cached(domain, app_build_id)
         config = GlobalAppConfig.by_app_id(domain, app.master_id)
+        version_info = _get_version_info(config)
```
Comment (mkangia) [inline corehq/apps/ota/views.py:321]: why the change in behaviour to send this each time? Looks like the mobile currently does not expect it in case it's not master app, so why send it? Just thinking about if it would change things on mobile.

Comment (snopoke) [reply corehq/apps/ota/views.py:321]: yea mean why remove the 'else'? Looking at the code the 'else' block was serving no function at all so I think it wasn't meant to be there.

Comment (mkangia) [reply corehq/apps/ota/views.py:321]: If i am reading correctly, it was meant to run if there was no exception. The config variable was unused so that can be removed in the old implementation.

Comment (snopoke) [reply corehq/apps/ota/views.py:321]: If you look at the original implementation it does update the request: https://github.com/dimagi/commcare-hq/blame/adc47bb934d4e9b75fef669d1ab55392adbceeed/corehq/apps/ota/views.py#L311

File: corehq/apps/app_manager/models.py (hunk 3 above)
Comment (snopoke) [reply corehq/apps/app_manager/models.py:5768]: > I think its fine if we want to just create it now, but would be nice to just delete it if that is the case when fetching the app?

I don't think this is necessary and it could create a lot of churn in that DB table which is more harmful

> And how do we avoid fetching the app in the request now? looks like we are just delaying it?

We only fetch it if there isn't a cache for the values.

Comment (mkangia) [reply corehq/apps/app_manager/models.py:5768]: Oh cool, so the cache `get_latest_app_version` returns that without needing us to load app,
thanks!

File: corehq/apps/ota/views.py (hunk 4 above)
Comment (mkangia) [reply corehq/apps/ota/views.py:321]: I see, looks like it fetched it but it never put it in the `info` that was returned?

Comment (snopoke) [reply corehq/apps/ota/views.py:321]: sorry, here's a better link: https://github.com/dimagi/commcare-hq/blob/8a323bef0069ef2ed944018485d8324e141d3437/corehq/apps/ota/views.py#L274

Comment (mkangia) [reply corehq/apps/ota/views.py:321]: oh cool.

**Comment by mkangia** [comment]
Build failures look related @snopoke

**Comment by sravfeyn** [comment]
Actually, is there any reason to not cache the whole view `corehq.apps.ota.views.heartbeat` for 24 hours?

**Comment by snopoke** [comment]
> Actually, is there any reason to not cache the whole view corehq.apps.ota.views.heartbeat for 24 hours? (except the update_user_reporting_data part)

Probably could do that. I'm not going to change it here since we need to get this out ASAP

**Comment by snopoke** [comment]
@mkangia @sravfeyn any more thoughts or can I get an approval to merge.

**Review by mkangia — APPROVED** [review]

**Comment by snopoke** [comment]
This has definitely made a difference but we're still seeing high network usage on couch from the profile based feature:
image
